### ISSUE TITLE
Composer TYPO3 commands error, when mutagen enabled

### ISSUE THREAD

**Issue by lukasdoerr** [drud/ddev#4152]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Run a Diagnostic and Paste Link Here

https://gist.github.com/lukasdoerr/be134a6c3b76eaa78ccb2ab98bdedc3d
I have also tried to fix it with this issue: https://github.com/drud/ddev/issues/3384

+ System Info (Because i use WSL-2)

OS Name:                   Microsoft Windows 10 Pro
OS Version:                10.0.19044 N/A Build 19044
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          Lukas
Registered Organization:
Original Install Date:     27/08/2022, 00:39:51
System Boot Time:          29/08/2022, 12:48:03
System Manufacturer:       To Be Filled By O.E.M.
System Model:              To Be Filled By O.E.M.
System Type:               x64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: AMD64 Family 23 Model 113 Stepping 0 AuthenticAMD ~3600 Mhz
System Locale:             en-gb;English (United Kingdom)
Input Locale:              de;German (Germany)
Time Zone:                 (UTC+01:00) Amsterdam, Berlin, Bern, Rome, Stockholm, Vienna
Total Physical Memory:     16.310 MB

### Current Behavior

When i want to update/install/dump-autoload via ddev composer command, every command failes due to missing permissions in typo3cms command.

So i want to dump the autoload from my ext. in TYPO3. i run: 
`ddev composer dump-autoload`
i get this back:
`lukas@DESKTOP-NKGIEV8:/mnt/d/DDEV/typo3test$ ddev composer dump-autoload
Generating autoload files
> typo3cms install:fixfolderstructure
sh: 1: typo3cms: Permission denied
Script typo3cms install:fixfolderstructure handling the typo3-cms-scripts event returned with error code 126
Script @typo3-cms-scripts was called via post-autoload-dump
composer [dump-autoload] failed, composer command failed: exit status 126. stderr=`

Also caching is very bad rn, i can delete the cache via TYPO3, disable caching.. but when i want to change values in my self-written extension, the values are still cached. Only if i stop ddev or restart it (Without mutagen) it flushes the cache and shows the new value.

I have to disable mutagen and restart ddev to be able to use my composer commands:

`ddev config --mutagen-enabled=false
You are reconfiguring the project at /mnt/d/DDEV/typo3test.
The existing configuration will be updated and replaced.
Generating AdditionalConfiguration.php file for database connection.
Configuration complete. You may now run 'ddev start'.
lukas@DESKTOP-NKGIEV8:/mnt/d/DDEV/typo3test$ ddev restart`

But the problem then is, that the page and TYPO3 loads so slow, i can get a coffee in that time.

### Expected Behavior

I should be able to install/update/flush extensions with mutagen enabled. And the cache should be flushed, when resetting cache via TYPO3 backend + ddev composer dump-autoload

### Steps To Reproduce

Using Windows Docker WSL-2

1. Create a DDEV TYPO3 v11(main) image
2. Install it via ddev composer
3. Enable mutagen
4. Try to command: ddev composer install whatever:extension or ddev composer dump-autoload

### Anything else?

Keep in mind, that it works flawless when mutagen is disabled. I have also tried it directly on linux (Ubuntu Server 22.04 LTS) but it is not happening there.

**Comment by rfay** [comment]
Shouldn't you be using `ddev composer require <extension>`  and not `ddev composer install <extension>` ? Please show the exact sequence of commands you're using and what happens. A gist on gist.github.com would be nice.

But I see "sh: 1: typo3cms: Permission denied", which means that typo3cms isn't executable, which probably means that you don't have your mode settings right in git and you checked it in. 

I think if you `ls -l vendor/bin` or `ddev exec ls -l vendor/bin` you'll see that the items there aren't executable. Can you show those?

**Comment by lukasdoerr** [comment]
Hey Randy, thanks for your quick answer! Of course i mean composer require, not install :D Sorry for that! But in this case, i can`t run any composer command, that runs something or makes folder changes (due to this folder permissions problem).

Here is an example for "ddev composer dump-autoload", when mutagen is OFF:
https://gist.github.com/lukasdoerr/25ef0bfa665ac0158caf4776fad4de5e

And here is an example of "ddev composer dump-autoload", when mutagen is ON:
https://gist.github.com/lukasdoerr/af8d674efc12475ea1005091749918b4

"ls -l vendor/bin" spits out this info:
https://gist.github.com/lukasdoerr/91c6c610d392eb35c06cc2240f70da47

with "ddev exec ls -l vendor/bin" it spits the same out, but it looks like they are not runnable (should i do chmod +x on them?)

When running a command manually in TYPO3 folder, it shows me this:

> lukas@DESKTOP-NKGIEV8:/mnt/d/DDEV/typo3test$ vendor/bin/typo3
> /usr/bin/env: ‘php’: No such file or directory

i just used this tutorial to install TYPO3 locally: https://typo3.com/blog/tutorial-get-a-local-typo3-v10-installation-with-no-effort

**Comment by rfay** [comment]
It will be interesting to compare the DDEV TYPO3 quickstart, https://ddev.readthedocs.io/en/latest/users/quickstart/#typo3-composer-build, which works fine, right? What's the difference do you think?

I think you may have discovered what's going on by noting that the items in vendor/bin are not executable inside the container. That will definitely break everything.

Has any of this been checked in, or did you build it all from scratch using the article?

**Comment by lukasdoerr** [comment]
What the hell, it works now!
I have used the command:
`$ ddev exec chmod +x -R vendor/bin`
and i can now use the ddev composer commands while mutagen is enabled!

Caching works now as well as expected.
Can you imagine, why my ddev project had permission problems?

**Comment by rfay** [comment]
I really want to know why the two approaches are different. 

BTW, mutagen isn't useful when working with Linux.... so you really didn't gain anything by turning it on in the first place.

Would you be willing to take the time to go through both the article approach and the DDEV doc quickstart approach step-by-step and see if you can recreate this situation and detect what the difference is?

**Comment by lukasdoerr** [comment]
I will, but i have not much time due to work. I will try it at the weekend and write my results here. Do i have to close it here, or do we keep it open?

**Comment by rfay** [comment]
Let's keep it open. I think the official DDEV TYPO3 installation instructions on typo3.com don't work either with mutagen, but I can't find them right now.

**Comment by lukasdoerr** [comment]
Okay one thing that itches me (but im not sure if its because of mutagen) is, that my variables in my contoller seem to be cached. I have disabled caching for the page, even with Cache from TYPO3 cleared it takes up to 30 seconds to see my new value. But i can`t really compare it to non-mutagen mode, because without mutagen the page loads as well like 30 seconds. Is there a way to disable caching from mutagen for specific folders to check if its because of it?

**Comment by rfay** [comment]
Oh, if you have a page load of 30s in WSL2 it means you have your project on the WIndows filesystem instead of having it in /home. You really, really don't want that. Can you show the `pwd` of your project?

**Comment by lukasdoerr** [comment]
Haha i deleted it, because i thought its because of my extension. But if you think it could be because of the windows filesystem, that could be the issue, because i can`t find out otherwise. It mounts to my D:\ on windows, so i can have PHP Intellisense for it:

/mnt/d/DDEV/typo3test

Is there a better way? I really need the typo3_src connections to have auto-fill and code completion.

**Comment by rfay** [comment]
You absolutely don't want to put it on your WIndows, it's a catastrophe. It's not worth using wsl2 in that case. It's even slower than wsl2. See https://ddev.readthedocs.io/en/latest/users/install/ddev-installation/#windows-wsl2

You need .5s page loads, not 30s page loads. 

The way you're doing it, there are 3 layers of network copying. Windows to WSL2 via a network share, then wsl2 into docker via mutagen in this case. It's horrible. You'll be amazed at the performance difference.

**Comment by rfay** [comment]
You don't say what IDE you're using... Both PhpStorm and vscode have fair solutions to this.

**Comment by lukasdoerr** [comment]
> You don't say what IDE you're using... Both PhpStorm and vscode have fair solutions to this.

Im using vscode mainly, what kind of solution do you mean?

**Comment by rfay** [comment]
vscode has a lovely setup, see https://ddev.com/ddev-local/ddev-wsl2-getting-started/

You just use `code .` inside wsl2 in your project directory. But please, please put your code in the /home directory where it's on a linux filesystem.

**Comment by lukasdoerr** [comment]
i will, thank you! Keeping you updated when im installing the other method!

**Comment by lukasdoerr** [comment]
I have successfully installed TYPO3 via the quickstart tutorial. The only thing what was missing, was the Additionalconfiguration for database. I have just added it manually and it works.

I have now switched to WSL-2 via Remote Explorer and it runs so fast even without mutagen enabled. Caching Problems are gone now.

My guess is: ddev has problems in my situation, when not using WSL-2 and even having to synchronize my project to the webserver from windows. Now it runs very smooth, but i have to get some routine, working with vscode Remote Management.

I also have to check, how i can work with Github Desktop in this case - but thats a story for another time.

Thank you for all your information. Im mostly sure both tutorials work the same.
